src/sql/workbench/contrib/query/browser/fileQueryResultsWriter.ts

```diff
@@ -0,0 +1,341 @@
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { hyperLinkFormatter, textFormatter } from 'sql/base/browser/ui/table/formatters';
+import { IQueryEditorConfiguration } from 'sql/platform/query/common/query';
+import { IGridDataProvider } from 'sql/workbench/services/query/common/gridDataProvider';
+import { IQueryMessage, ResultSetSummary, IQueryResultsWriter } from 'sql/workbench/services/query/common/query';
+import QueryRunner, { QueryGridDataProvider } from 'sql/workbench/services/query/common/queryRunner';
+import { asArray } from 'vs/base/common/arrays';
+import { CancellationToken } from 'vs/base/common/cancellation';
+import { Disposable } from 'vs/base/common/lifecycle';
+import { isArray } from 'vs/base/common/types';
+import { isWindows } from 'vs/base/common/platform';
+import { formatDocumentWithSelectedProvider, FormattingMode } from 'vs/editor/contrib/format/format';
+import * as nls from 'vs/nls';
+import { IConfigurationService } from 'vs/platform/configuration/common/configuration';
+import { IInstantiationService } from 'vs/platform/instantiation/common/instantiation';
+import { ILogService } from 'vs/platform/log/common/log';
+import { Progress } from 'vs/platform/progress/common/progress';
+import { IEditorService } from 'vs/workbench/services/editor/common/editorService';
+import { IUntitledTextEditorService } from 'vs/workbench/services/untitled/common/untitledTextEditorService';
+
+export class FileQueryResultsWriter implements IQueryResultsWriter {
+	private messages: Array<IQueryMessage> = [];
+	private formattedQueryResults: Array<string> = [];
+	private tables: Array<Table<any>> = [];
+	private numQueriesToFormat: number = 0;
+	private queryContainsError: boolean = false;
+	private closingMessageIncluded: boolean = false;
+	private hasCreatedResultsFile: boolean = false;
+
+	private readonly startedExecutingQueryMessage: string;
+	private readonly totalExecutionTimeMessage: string;
+	private readonly rowAffectedMessage: string;
+	private readonly rowsAffectedMessage: string;
+	private readonly newLineEscapeSequence: string;
+
+	constructor(
+		private runner: QueryRunner,
+		@IInstantiationService private instantiationService: IInstantiationService,
+		@IUntitledTextEditorService private readonly untitledEditorService: IUntitledTextEditorService,
+		@IEditorService private readonly editorService: IEditorService,
+		@ILogService private readonly logService: ILogService,
+		@IConfigurationService private configurationService: IConfigurationService
+	) {
+		this.startedExecutingQueryMessage = nls.localize('query.message.startingExecutionQuery', 'Started executing query');
+		this.totalExecutionTimeMessage = nls.localize('query.message.totalExecutionTime', 'Total execution time');
+		this.rowAffectedMessage = nls.localize('query.message.rowAffected', 'row affected');
```

Comment:
You shouldn't create a new one for each instance - you can just create a local const at the top level of the module (outside the class) and use it elsewhere in the file (like how we do for the "localized strings" files in many places)

But also, for messages like `Total execution time` that are added in ADS we shouldn't be making our own strings here. If a translation mistake happens (such as the translators using different phrasing for incomplete sentences like this) then we could easily end up with strings that don't match what the actual message has. And in general doing string comparisons like this isn't a great pattern

I would instead recommend that you find some way to mark the actual message objects themselves with a flag that tells it whether to output or not. Although I'm curious what logic you used to determine what was/wasn't included in the output file - did you have some list you were using for determining that?

Reply:
I didn't have a list to determine what to include, so I used SSMS to help determine that. I found that "row(s) affected" message is coming from STS, so I can add a flag to the data transfer object (DTO) that is being sent from STS to ADS. The "total execution time message" and "started executing query" message both come from the query runner, so adding a flag there should be simple.

Where does ADS handle mapping data from a ResultMessage object to an IQueryMessage object?

Reply:
You mean IResultMessage? There isn't a "mapping" per-se, ADS hooks into the onMessage notification from STS and the query runner then sends that to the output panel. The ResultMessage object is turned into JSON for transferring over the JSON RPC layer.

So all you'd need todo is  update the interface/class in the places it's defined and then have STS send the message with the new flag : 

https://github.com/microsoft/azuredatastudio/blob/main/src/sql/azdata.d.ts#L925
https://github.com/microsoft/azuredatastudio/blob/main/src/sql/workbench/services/query/common/query.ts#L53

https://github.com/microsoft/sqlops-dataprotocolclient/blob/main/src/types.ts#L507

https://github.com/Microsoft/sqltoolsservice/blob/main/src/Microsoft.SqlTools.ServiceLayer/QueryExecution/Contracts/ResultMessage.cs